model_discretization_api.py

```diff
@@ -0,0 +1,209 @@
+# like:
+source = Box2D(1, 0, 0, 1, 1)
+psf = Gaussian2DKernel(1)
+convolved_model = ConvolvedModel(source, psf, mode='oversample', factor=10)
+
+# mode will be used both for convolution and later integration
+# if a different mode is desired for convolution it should be passed in the psf
+
+# support of 1D models
+source = Lorentz1D(1, 0, 0, 1, 1)
+psf = Gaussian1DKernel(1)
+convolved_model = ConvolvedModel(source, psf)
+
+# should there be a Convolved1DModel and Convolved2DModel?
+# should this be an extension of the CompoundModel class?
+
+# later we can think about using any model as convolution kernels
+# and even implement the analytical solution for the common Gaussian case
+# or other cases where analytical solutions are known
+source = Gaussian2D(1, 0, 0, 1, 1)
+psf = Gaussian2D(1, 0, 0, 1, 1)
+convolved_model = ConvolvedModel(source, psf)
+
+
+# the convolution method can be set via string, to choose between the astropy.convolution
+# methods
+convolved_model = ConvolvedModel(source, psf, method='fft')
+convolved_model = ConvolvedModel(source, psf, method='standard')
+
+# or by passing a function that should be used. This is mainly nececessary, because
+# the performance of the astropy convolution algorithms is rather weak...
+convolved_model = ConvolvedModel(source, psf, method=scipy.signal.fftconvolve)
+
+# recommendation would be to use scipy.signal.fftconvolve which is very fast
+
+# additional keyword arguments can be passed to the convolution function
+convolved_model = ConvolvedModel(source, psf, method='standard', mode='oversample')
+
+# the convolved_model can be fit 
+
+# Further notes:
+# 1. Implementation of this class should be straight forward and not much effort
+# 2. Testing could be set up using Gaussian models, where the analytical solution is
+# known
+# 3. Later one could think about assigning the corresponding fourier transforms to analytical
+# models and use this for the convolution, which will probably have a better performance.
+# Source model and PSF are evaluated analytically in Fourier space, mutliplied and transformed
+# back to normal space.
+# 4. As convolution is a linear operation the derivative of a convolution behaves like 
+# http://en.wikipedia.org/wiki/Convolution#Differentiation this relation can be used
+# to implement `fit_deriv` methods for convolved models.
+
+
+#########################
+# Model integration API #
+#########################
+
+import numpy as np
+
+# to avoid bias, when the scale of the model is similar to the size of the pixels,
+# models have to be integrated over pixels, for this purpose models could define
+# an integrate function, that takes upper and lower bounds of the integration
+# (see also https://github.com/astropy/astropy/issues/1586)
+model = Gaussian1D(1, 0, 1)
+model.integrate(0, np.inf)
+
+# it should also work with arrays of bounds of course
+x_lo = np.arange(-10, 11) - 0.5
+x_hi = np.arange(-10, 11) + 0.5
+model.integrate((x_lo, x_hi))
```

Comment:
For broad-band SED integration integration there's (at least) two things to consider to implement it well:
* The numerical integration should be done with `x = log(energy)` and `y = log(flux)` axes.
* To compute the integral count flux in an energy band one has to integrate `y = log(flux)`, to get the integral energy flux one has to integrate `y = log(flux * energy ** 2)` I think this is similar to the use case @eteq mentioned of adding factors of r or r^2 to polar and spherical coordinate integrals, i.e. there is a coordinate transform involved, that's why I was mentioning the SED use case here.

Examples of SED integration Python code is [here](https://github.com/zblz/naima/blob/master/naima/utils.py#L24), [here](https://github.com/jbudynk/sedstacker/blob/master/sedstacker/sed.py#L429) and [here](https://github.com/gammapy/gammapy/blob/88b22048a69ea5941cbf6647ab3464218edce66b/gammapy/spectrum/models.py#L82).

Like I said, it would be nice if the Astropy model integration methods supported such computation somehow, but I don't have a suggestion how it should be implemented. So my suggestion here would be: if someone posts a concrete idea for this feature here within ~ the next week, let's try to flesh it out, otherwise punt on it as possible future work or something to be done elsewhere on top of astropy.modeling.